src/main/scala/cromwell/engine/WorkflowActor.scala

```diff
@@ -1,138 +1,91 @@
-                                 backend: Backend) extends Actor {
+  sealed trait WorkflowFailure
+  case object NoFailureMessage extends WorkflowFailure
+  case class FailureMessage(msg: String) extends WorkflowFailure with WorkflowActorMessage
+}
 
-  private val log = Logging(context.system, this)
+case class WorkflowActor(id: WorkflowId,
+                         namespace: WdlNamespace,
+                         actualInputs: WorkflowCoercedInputs,
+                         backend: Backend) extends LoggingFSM[WorkflowState, WorkflowFailure] {
+  implicit val timeout = Timeout(5 seconds)
 
   private val storeActor = context.actorOf(StoreActor.props(namespace, actualInputs))
 
-  def receive: Receive = unstarted
+  startWith(WorkflowSubmitted, NoFailureMessage)
 
-  /** Unstarted `Receive` state that only responds to `Start` messages. */
-  def unstarted: Receive = LoggingReceive {
-    case WorkflowActor.Start => handleInitialStart()
-    case bad => fail(bad, "unstarted")
+  when(WorkflowSubmitted) {
+    case Event(WorkflowActor.Start, NoFailureMessage) => goto(WorkflowRunning)
   }
 
-  /** Started `Receive` state. */
-  def started: Receive = LoggingReceive {
-    case CallActor.Started(call) =>
+  when(WorkflowRunning) {
+    case Event(CallActor.Started(call), NoFailureMessage) =>
       storeActor ! UpdateStatus(call, ExecutionStatus.Running)
-
-    case CallActor.Completed(completedCall, callOutputs) =>
-      handleCallCompleted(completedCall, callOutputs)
-
-    case CallActor.Failed(call, failure) =>
-      storeActor ! UpdateStatus(call, ExecutionStatus.Failed)
-      context.parent ! WorkflowActor.Failed(failure)
-
-    case WorkflowActor.GetOutputs =>
-      (storeActor ? StoreActor.GetOutputs) pipeTo sender
-
-    case bad => fail(bad, "started")
-  }
-
-  /** Message the parent actor with `Failed` with an appropriate message. */
-  private def fail(badMessage: Any, stateName: String): Unit = {
-    val diagnostic = s"Unexpected message $badMessage received when in $stateName state."
-    log.error(diagnostic)
-    context.parent ! WorkflowActor.Failed(diagnostic)
-  }
-
-  /**
-   * Performs the following steps:
-   * <ol>
-   *   <li>Finds all runnable calls.</li>
-   *   <li>Copies outputs from prerequisite calls to the inputs of the runnable calls.</li>
-   *   <li>Starts a call actor for each runnable call.</li>
-   * </ol>
-   */
-  private def startRunnableCalls(): Unit = {
-    for {
-      runnableCalls <- (storeActor ? FindRunnableCalls).mapTo[Iterable[Call]]
-    } yield {
+      stay()
+    case Event(CallActor.Completed(completedCall, callOutputs), NoFailureMessage) =>
+      storeActor ! CallCompleted(completedCall, callOutputs)
+      stay()
+    case Event(StoreActor.RunnableCalls(runnableCalls), NoFailureMessage) =>
       if (runnableCalls.nonEmpty) {
         log.info("Starting calls: " + runnableCalls.map {_.name}.toSeq.sorted.mkString(", "))
       }
       runnableCalls foreach startCallActor
```

Comment:
Oops, forgot the specific to this one angle - I believe I wrote it when I was in a more kowtowing mood ;) Actually my thought was that since both statements were side effects you might as well put them here.